File: web/src/__tests__/dsl/runtime/NoteMutation.test.ts
import { describe, it, expect, vi } from 'vitest'
import { Lexer } from '../../../dsl/language/Lexer'
import { Parser } from '../../../dsl/language/Parser'
import { Executor } from '../../../dsl/runtime/Executor'
import { Environment } from '../../../dsl/runtime/Environment'
import type { DslValue, NumberVal, StringVal, NoteVal } from '../../../dsl/runtime/DslValue'
import type { NoteOperations } from '../../../dsl/runtime/NoteOperations'
import type { Note } from '../../../data/Note'

function execute(source: string, env?: Environment): DslValue {
  const tokens = new Lexer(source).tokenize()
  const directive = new Parser(tokens, source).parseDirective()
  return new Executor().execute(directive, env ?? Environment.create())
}

/**
 * Execute a directive that contains async operations (new, maybe_new, append, property set).
 * The executor returns a Promise wrapped as a DslValue for async builtins.
 */
async function executeAsync(source: string, env: Environment): Promise<DslValue> {
  const result = execute(source, env)
  // If the result is a Promise (from async builtin), await it
  if (result && typeof (result as unknown as Promise<DslValue>).then === 'function') {
    return (result as unknown as Promise<DslValue>)
  }
  return result
}

function makeNote(overrides: Partial<Note> & { id: string }): Note {
  return {
    userId: 'user1',
    parentNoteId: null,
    content: overrides.content ?? '',
    createdAt: null,
    updatedAt: null,
    lastAccessedAt: null,
    tags: [],
    containedNotes: [],
    state: null,
    path: overrides.path ?? '',
    rootNoteId: null,
    showCompleted: true,
    ...overrides,
  }
}

function createMockNoteOperations(notes: Note[] = []): NoteOperations {
  const noteMap = new Map(notes.map((n) => [n.id, { ...n }]))

  return {
    createNote: vi.fn(async (path: string, content: string): Promise<Note> => {
      const note = makeNote({ id: `new-${Date.now()}`, path, content })
      noteMap.set(note.id, note)
      return note
    }),
    getNoteById: vi.fn(async (noteId: string): Promise<Note | null> => {
      return noteMap.get(noteId) ?? null
    }),
    findByPath: vi.fn(async (path: string): Promise<Note | null> => {
      for (const note of noteMap.values()) {
        if (note.path === path) return note
      }
      return null
    }),
    noteExistsAtPath: vi.fn(async (path: string): Promise<boolean> => {
      for (const note of noteMap.values()) {
        if (note.path === path) return true
      }
      return false
    }),
    updatePath: vi.fn(async (noteId: string, newPath: string): Promise<Note> => {
      const note = noteMap.get(noteId)
      if (!note) throw new Error(`Note not found: ${noteId}`)
      const updated = { ...note, path: newPath }
      noteMap.set(noteId, updated)
      return updated
    }),
    updateContent: vi.fn(async (noteId: string, newContent: string): Promise<Note> => {
      const note = noteMap.get(noteId)
      if (!note) throw new Error(`Note not found: ${noteId}`)
      const updated = { ...note, content: newContent }
      noteMap.set(noteId, updated)
      return updated
    }),
    appendToNote: vi.fn(async (noteId: string, text: string): Promise<Note> => {
      const note = noteMap.get(noteId)
      if (!note) throw new Error(`Note not found: ${noteId}`)
      const updated = { ...note, content: note.content + '\n' + text }
      noteMap.set(noteId, updated)
      return updated
    }),
  }
}

describe('NoteMutation', () => {
  describe('variable assignment', () => {
    it('should assign simple value', () => {
      const result = execute('[x: 42; x]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(42)
    })

    it('should assign string value', () => {
      const result = execute('[x: "hello"; x]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('hello')
    })

    it('should assign variable referencing another variable', () => {
      const result = execute('[x: 10; y: x; y]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(10)
    })

    it('should handle multiple variable assignments', () => {
      const result = execute('[a: 1; b: 2; add(a, b)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(3)
    })

    it('should throw on undefined variable', () => {
      expect(() => execute('[undefined_var]')).toThrow()
    })
  })

  describe('semicolons', () => {
    it('should execute statements separated by semicolons', () => {
      const result = execute('[x: 1; y: 2; add(x, y)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(3)
    })

    it('should return last statement value', () => {
      const result = execute('[1; 2; 3]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(3)
    })
  })

  describe('property assignment - path', () => {
    it('should assign path property', () => {
      const note = makeNote({ id: 'n1', path: 'old/path', content: 'Test' })
      const ops = createMockNoteOperations([note])
      const env = Environment.withAll([note], note, ops)

      // setProperty is async but the executor fires-and-forgets it;
      // we just verify the mock was called.
      execute('[.path: "new/path"]', env)

      expect(ops.updatePath).toHaveBeenCalledWith('n1', 'new/path')
    })
  })

  describe('property assignment - name', () => {
    it('should assign name property', () => {
      const note = makeNote({ id: 'n1', path: 'test', content: 'Old Name\nLine 2' })
      const ops = createMockNoteOperations([note])
      const env = Environment.withAll([note], note, ops)

      execute('[.name: "New Name"]', env)

      expect(ops.getNoteById).toHaveBeenCalledWith('n1')
    })
  })

  describe('method calls - append', () => {
    it('should call append method', () => {
      const note = makeNote({ id: 'n1', path: 'test', content: 'Initial' })
      const ops = createMockNoteOperations([note])
      const env = Environment.withAll([note], note, ops)

      execute('[.append("Added line")]', env)

      expect(ops.appendToNote).toHaveBeenCalledWith('n1', 'Added line')
    })
  })

  describe('new function', () => {
    it('should create new note', async () => {
      const ops = createMockNoteOperations([])
      const env = Environment.withNoteOperations(ops)

      // new() is async, so the executor returns a Promise as a DslValue
      await executeAsync('[new(path: "test/new")]', env)

      expect(ops.noteExistsAtPath).toHaveBeenCalledWith('test/new')
      expect(ops.createNote).toHaveBeenCalledWith('test/new', '')
    })

    it('should create new note with content', async () => {
      const ops = createMockNoteOperations([])
      const env = Environment.withNoteOperations(ops)

      await executeAsync('[new(path: "test/new", content: "Hello")]', env)

      expect(ops.createNote).toHaveBeenCalledWith('test/new', 'Hello')
    })

    it('should throw when note already exists', async () => {
      const existing = makeNote({ id: 'n1', path: 'test/existing', content: 'Existing' })
      const ops = createMockNoteOperations([existing])
      const env = Environment.withNoteOperations(ops)

      await expect(executeAsync('[new(path: "test/existing")]', env)).rejects.toThrow()
    })

    it('should reject without path argument', async () => {
      const ops = createMockNoteOperations([])
      const env = Environment.withNoteOperations(ops)

      // new() is async; missing path throws inside the async fn, resulting in rejection
      await expect(executeAsync('[new()]', env)).rejects.toThrow(/path/)
    })
  })

  describe('maybe_new function', () => {
    it('should create note when it does not exist', async () => {
      const ops = createMockNoteOperations([])
      const env = Environment.withNoteOperations(ops)

      await executeAsync('[maybe_new(path: "test/new")]', env)

      expect(ops.findByPath).toHaveBeenCalledWith('test/new')
      expect(ops.createNote).toHaveBeenCalledWith('test/new', '')
    })

    it('should return existing note when it exists', async () => {
      const existing = makeNote({ id: 'n1', path: 'test/existing', content: 'Existing' })
      const ops = createMockNoteOperations([existing])
      const env = Environment.withNoteOperations(ops)

      const result = await executeAsync('[maybe_new(path: "test/existing")]', env)

      expect(ops.findByPath).toHaveBeenCalledWith('test/existing')
      expect(result.kind).toBe('NoteVal')
      expect((result as NoteVal).note.id).toBe('n1')
    })

    it('should reject without path argument', async () => {
      const ops = createMockNoteOperations([])
      const env = Environment.withNoteOperations(ops)

      await expect(executeAsync('[maybe_new()]', env)).rejects.toThrow(/path/)
    })
  })

  describe('combined functionality', () => {
    it('should assign variable and use it in expression', () => {
      const result = execute('[x: 5; y: 3; mul(x, y)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(15)
    })

    it('should assign computed value to variable', () => {
      const result = execute('[x: add(2, 3); mul(x, x)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(25)
    })

    it('should chain variable assignments', () => {
      const result = execute('[a: 1; b: add(a, 1); c: add(b, 1); c]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(3)
    })

    it('should use note variables in expressions', () => {
      const note = makeNote({ id: 'n1', path: 'tasks/daily', content: 'My Task' })
      const env = Environment.withNotesAndCurrentNote([note], note)

      const result = execute('[n: .; n.path]', env)
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('tasks/daily')
    })

    it('should access note name via property', () => {
      const note = makeNote({ id: 'n1', path: 'test', content: 'Note Title\nBody' })
      const env = Environment.withNotesAndCurrentNote([note], note)

      const result = execute('[.name]', env)
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('Note Title')
    })

    it('should access note id via property', () => {
      const note = makeNote({ id: 'n1', path: 'test', content: 'Content' })
      const env = Environment.withNotesAndCurrentNote([note], note)

      const result = execute('[.id]', env)
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('n1')
    })
  })
})

File: app/src/test/java/org/alkaline/taskbrain/dsl/runtime/NoteMutationTest.kt
package org.alkaline.taskbrain.dsl.runtime

import com.google.firebase.Timestamp
import org.alkaline.taskbrain.dsl.runtime.values.DslValue
import org.alkaline.taskbrain.dsl.runtime.values.ListVal
import org.alkaline.taskbrain.dsl.runtime.values.NoteVal
import org.alkaline.taskbrain.dsl.runtime.values.NumberVal
import org.alkaline.taskbrain.data.Note
import org.alkaline.taskbrain.dsl.language.Lexer
import org.alkaline.taskbrain.dsl.language.Parser
import org.alkaline.taskbrain.testutil.MockNoteOperations
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.util.Date

/**
 * Tests for Note Mutation functionality.
 *
 * Tests cover:
 * - Variable assignment: [x: 5]
 * - Statement separation: [x: 5; y: 10; add(x, y)]
 * - Property assignment: [.path: "value"]
 * - Method calls: [.append("text")]
 * - new() function
 * - maybe_new() function
 */
class NoteMutationTest {

    private lateinit var executor: Executor
    private lateinit var mockOps: MockNoteOperations

    @Before
    fun setUp() {
        executor = Executor()
        mockOps = MockNoteOperations()
    }

    private fun execute(
        source: String,
        notes: List<Note>? = null,
        currentNote: Note? = null,
        noteOps: NoteOperations? = mockOps
    ): DslValue {
        val tokens = Lexer(source).tokenize()
        val directive = Parser(tokens, source).parseDirective()
        val env = Environment(NoteContext(
            notes = notes,
            currentNote = currentNote,
            noteOperations = noteOps
        ))
        return executor.execute(directive, env)
    }

    // region Sample notes

    private val testNote = Note(
        id = "test-note-1",
        userId = "user1",
        path = "journal/2026-01-25",
        content = "Test content",
        createdAt = Timestamp(Date())
    )

    // endregion

    // region Variable assignment tests

    @Test
    fun `variable assignment returns assigned value`() {
        val result = execute("[x: 5]")

        assertTrue(result is NumberVal)
        assertEquals(5.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `variable can be referenced after assignment`() {
        val result = execute("[x: 42; x]")

        assertTrue(result is NumberVal)
        assertEquals(42.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `multiple variables can be defined and used`() {
        val result = execute("[x: 10; y: 20; add(x, y)]")

        assertTrue(result is NumberVal)
        assertEquals(30.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `undefined variable throws error`() {
        try {
            execute("[undefinedVar]")
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            // Error message says "Unknown function or variable" when not found
            assertTrue(e.message?.contains("Unknown function or variable") == true)
        }
    }

    // endregion

    // region Statement separation tests

    @Test
    fun `semicolon separates statements`() {
        val result = execute("[1; 2; 3]")

        // Returns value of last statement
        assertTrue(result is NumberVal)
        assertEquals(3.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `statements execute in order`() {
        val result = execute("[a: 1; a: add(a, 1); a: add(a, 1); a]")

        assertTrue(result is NumberVal)
        assertEquals(3.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `single statement without semicolon works`() {
        val result = execute("[add(1, 2)]")

        assertTrue(result is NumberVal)
        assertEquals(3.0, (result as NumberVal).value, 0.0)
    }

    // endregion

    // region Property assignment tests

    @Test
    fun `assign path to current note`() {
        mockOps.addNote(testNote)

        execute("[.path: \"new/path\"]", currentNote = testNote)

        assertEquals(1, mockOps.updatedPaths.size)
        assertEquals(testNote.id to "new/path", mockOps.updatedPaths[0])
    }

    @Test
    fun `assign name to current note`() {
        mockOps.addNote(testNote)

        execute("[.name: \"New Title\"]", currentNote = testNote)

        assertEquals(1, mockOps.updatedContents.size)
        assertEquals(testNote.id, mockOps.updatedContents[0].first)
        assertEquals("New Title", mockOps.updatedContents[0].second)
    }

    @Test
    fun `assign name preserves rest of content`() {
        val noteWithMultipleLines = testNote.copy(content = "Old Title\nLine 2\nLine 3")
        mockOps.addNote(noteWithMultipleLines)

        execute("[.name: \"New Title\"]", currentNote = noteWithMultipleLines)

        assertEquals(1, mockOps.updatedContents.size)
        assertEquals("New Title\nLine 2\nLine 3", mockOps.updatedContents[0].second)
    }

    @Test
    fun `cannot assign to read-only property`() {
        mockOps.addNote(testNote)

        try {
            execute("[.id: \"new-id\"]", currentNote = testNote)
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("read-only") == true)
        }
    }

    @Test
    fun `property assignment without note operations throws error`() {
        try {
            execute("[.path: \"new/path\"]", currentNote = testNote, noteOps = null)
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("note operations not available") == true)
        }
    }

    // endregion

    // region Method call tests

    @Test
    fun `append method adds text to note`() {
        mockOps.addNote(testNote)

        val result = execute("[.append(\"New line\")]", currentNote = testNote)

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.appendedTexts.size)
        assertEquals(testNote.id to "New line", mockOps.appendedTexts[0])
    }

    @Test
    fun `append returns updated note for chaining`() {
        mockOps.addNote(testNote)

        val result = execute("[.append(\"Line 1\"); .append(\"Line 2\")]", currentNote = testNote)

        assertTrue(result is NoteVal)
        assertEquals(2, mockOps.appendedTexts.size)
    }

    @Test
    fun `append with variable text`() {
        mockOps.addNote(testNote)

        execute("[msg: \"Hello\"; .append(msg)]", currentNote = testNote)

        assertEquals(1, mockOps.appendedTexts.size)
        assertEquals("Hello", mockOps.appendedTexts[0].second)
    }

    @Test
    fun `append without note operations throws error`() {
        try {
            execute("[.append(\"text\")]", currentNote = testNote, noteOps = null)
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("note operations not available") == true)
        }
    }

    @Test
    fun `unknown method throws error`() {
        mockOps.addNote(testNote)

        try {
            execute("[.unknownMethod()]", currentNote = testNote)
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("Unknown method") == true)
        }
    }

    // endregion

    // region new() function tests

    @Test
    fun `new creates note at path`() {
        val result = execute("[new(path: \"test/path\", content: \"Hello\")]")

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.createdNotes.size)
        assertEquals("test/path", mockOps.createdNotes[0].path)
        assertEquals("Hello", mockOps.createdNotes[0].content)
    }

    @Test
    fun `new with empty content`() {
        val result = execute("[new(path: \"test/path\")]")

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.createdNotes.size)
        assertEquals("", mockOps.createdNotes[0].content)
    }

    @Test
    fun `new throws error when path exists`() {
        mockOps.addNote(Note(id = "existing", path = "test/path", content = ""))

        try {
            execute("[new(path: \"test/path\")]")
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("already exists") == true)
        }
    }

    @Test
    fun `new requires path argument`() {
        try {
            execute("[new(content: \"Hello\")]")
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("requires a 'path' argument") == true)
        }
    }

    @Test
    fun `new without note operations throws error`() {
        try {
            execute("[new(path: \"test/path\")]", noteOps = null)
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("requires note operations") == true)
        }
    }

    // endregion

    // region maybe_new() function tests

    @Test
    fun `maybe_new returns existing note`() {
        val existingNote = Note(id = "existing", userId = "user1", path = "test/path", content = "Existing content")
        mockOps.addNote(existingNote)

        val result = execute("[maybe_new(path: \"test/path\", maybe_content: \"New content\")]")

        assertTrue(result is NoteVal)
        assertEquals("existing", (result as NoteVal).note.id)
        assertEquals(0, mockOps.createdNotes.size)  // Should not create
    }

    @Test
    fun `maybe_new creates note when not exists`() {
        val result = execute("[maybe_new(path: \"test/path\", maybe_content: \"New content\")]")

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.createdNotes.size)
        assertEquals("test/path", mockOps.createdNotes[0].path)
        assertEquals("New content", mockOps.createdNotes[0].content)
    }

    @Test
    fun `maybe_new without maybe_content creates empty note`() {
        val result = execute("[maybe_new(path: \"test/path\")]")

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.createdNotes.size)
        assertEquals("", mockOps.createdNotes[0].content)
    }

    @Test
    fun `maybe_new requires path argument`() {
        try {
            execute("[maybe_new(maybe_content: \"Hello\")]")
            fail("Expected ExecutionException")
        } catch (e: ExecutionException) {
            assertTrue(e.message?.contains("requires a 'path' argument") == true)
        }
    }

    @Test
    fun `maybe_new is idempotent`() {
        execute("[maybe_new(path: \"test/path\", maybe_content: \"First call\")]")
        execute("[maybe_new(path: \"test/path\", maybe_content: \"Second call\")]")

        assertEquals(1, mockOps.createdNotes.size)
        assertEquals("First call", mockOps.createdNotes[0].content)
    }

    // endregion

    // region Combined functionality tests

    @Test
    fun `complex workflow with variables and mutations`() {
        mockOps.addNote(testNote)

        // Use a simpler workflow that doesn't depend on string() function
        val result = execute(
            "[msg: \"Test message\"; .append(msg)]",
            currentNote = testNote
        )

        assertTrue(result is NoteVal)
        assertEquals(1, mockOps.appendedTexts.size)
        assertEquals("Test message", mockOps.appendedTexts[0].second)
    }

    @Test
    fun `find notes and access properties`() {
        val notes = listOf(
            Note(id = "n1", userId = "u1", path = "2026-01-25", content = "Journal 1"),
            Note(id = "n2", userId = "u1", path = "2026-01-26", content = "Journal 2")
        )

        val result = execute(
            "[find(path: \"2026-01-25\")]",
            notes = notes
        )

        assertTrue(result is ListVal)
        assertEquals(1, (result as ListVal).size)
    }

    // endregion
}
